01-getting-started/projects/final/app/src/main/java/com/raywenderlich/podplay/viewmodel/PodcastViewModel.kt

```diff
@@ -0,0 +1,163 @@
+      activePodcast = podcast
+      podcastToSummaryView(podcast)
+    }
+  }
+
+  suspend fun getPodcast(podcastSummaryViewData: PodcastSummaryViewData) {
+    podcastSummaryViewData.feedUrl?.let { url ->
+      var podcast: Podcast?
+      // TODO 1
+      podcast = podcastRepo?.getPodcast(url)
+
+      podcast?.let {
+        // TODO 2
+        it.feedTitle = podcastSummaryViewData.name ?: ""
+        it.imageUrl = podcastSummaryViewData.imageUrl ?: ""
+        _podcastLiveData.value = podcastToPodcastView(it)
+        activePodcast = it
+      } ?: run {
+        _podcastLiveData.value = null
+      }
+    } ?: run {
+      _podcastLiveData.value = null
+    }
+  }
+
+  fun getPodcasts(): LiveData<List<PodcastSummaryViewData>>? {
+    val repo = podcastRepo ?: return null
+    if (livePodcastSummaryData == null) {
+      val liveData = repo.getAll()
+      livePodcastSummaryData = Transformations.map(liveData) { podcastList ->
+        podcastList.map { podcast ->
+          podcastToSummaryView(podcast)
+        }
+      }
+    }
+
+    return livePodcastSummaryData
+  }
+
+  fun saveActivePodcast() {
+    val repo = podcastRepo ?: return
+    activePodcast?.let {
+      repo.save(it)
+    }
+  }
+
+  private fun podcastToPodcastView(podcast: Podcast): PodcastViewData {
+    return PodcastViewData(
+        podcast.id != null,
+        podcast.feedTitle,
+        podcast.feedUrl,
+        podcast.feedDesc,
+        podcast.imageUrl,
+        episodesToEpisodesView(podcast.episodes)
+    )
+  }
+
+  private fun podcastToSummaryView(podcast: Podcast):
+      PodcastSummaryViewData {
+    return PodcastSummaryViewData(
+        podcast.feedTitle,
+        DateUtils.dateToShortDate(podcast.lastUpdated),
+        podcast.imageUrl,
+        podcast.feedUrl)
+  }
+
+  private fun episodesToEpisodesView(episodes: List<Episode>): List<EpisodeViewData> {
+    return episodes.map {
+      val isVideo = it.mimeType.startsWith("video")
+      EpisodeViewData(it.guid, it.title, it.description, it.mediaUrl, it.releaseDate, it.duration, isVideo)
+    }
+  }
+
+  fun deleteActivePodcast() {
+    val repo = podcastRepo ?: return
+    activePodcast?.let {
+      repo.delete(it)
+    }
+  }
+
+  data class PodcastViewData(var subscribed: Boolean = false,
+                             var feedTitle: String? = "",
+                             var feedUrl: String? = "",
+                             var feedDesc: String? = "",
+                             var imageUrl: String? = "",
+                             var episodes: List<EpisodeViewData>)
+
+  data class EpisodeViewData(var guid: String? = "",
+                             var title: String? = "",
+                             var description: String? = "",
+                             var mediaUrl: String? = "",
+                             var releaseDate: Date? = null,
+                             var duration: String? = "",
+                             var isVideo: Boolean = false)
+}
```

Comment:
I'd be super happy if we can refactor (simplify) this a bit. I know I'm asking too much, but I see a lot of room for improvement. I'll have some free time next week, maybe I can jump in to help you guys with this. Let me know!